### PR TITLE
Subsetting: prune redundant format 12 cmaps

### PR DESCRIPTION
**Pull request by madig**
Behdad was prescient at https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/subset/__init__.py#L2228-L2231.

My idea: 

1. Delete a `platformID="0" platEncID="4" format="12" language="X"` (Unicode platform, Unicode full repertoire) cmap if there is a `platformID="0" platEncID="3" language="X"` (Unicode BMP only) cmap that contains the same Unicode set after subsetting.
2. Delete a `platformID="3" platEncID="10" format="12" language="X"` (Windows platform, Unicode full repertoire) cmap if there is a `platformID="3" platEncID="1" language="X"` (Unicode BMP) cmap that contains the same Unicode set after subsetting.

### CODE REVIEW

File: Lib/fontTools/subset/__init__.py [hunk 1]
```diff
@@ -2207,28 +2207,58 @@ def prune_pre_subset(self, font, options):
...
+
 		if t.format == 14:
 			# TODO(behdad) We drop all the default-UVS mappings
 			# for glyphs_requested.  So it's the caller's responsibility to make
 			# sure those are included.
 			t.uvsDict = {v:[(u,g) for u,g in l
-					      if g in s.glyphs_requested or u in s.unicodes_requested]
-				     for v,l in t.uvsDict.items()}
+						  if g in s.glyphs_requested or u in s.unicodes_requested]
+					 for v,l in t.uvsDict.items()}
```
Comment (anthrotype) [inline Lib/fontTools/subset/__init__.py]: is the whitespace change only cosmetic? (one day we'll blacken everything so I wouldn't bother)

File: Lib/fontTools/subset/__init__.py [hunk 2]
```diff
@@ -2207,28 +2207,58 @@ def prune_pre_subset(self, font, options):
 @_add_method(ttLib.getTableClass('cmap'))
 def subset_glyphs(self, s):
 	s.glyphs = None # We use s.glyphs_requested and s.unicodes_requested only
+
+	tables_format12 = []
+	table_plat0_enc3 = {}
+	table_plat3_enc1 = {}
```
Comment (anthrotype) [inline Lib/fontTools/subset/__init__.py:2213]: pls add comment describing what `plat` and `enc` number means

File: Lib/fontTools/subset/__init__.py [hunk 3]
```diff
@@ -2207,28 +2207,58 @@ def prune_pre_subset(self, font, options):
...
+		):
+			t.cmap.clear()
+
 	self.tables = [t for t in self.tables
 			 if (t.cmap if t.format != 14 else t.uvsDict)]
 	self.numSubTables = len(self.tables)
-	# TODO(behdad) Convert formats when needed.
-	# In particular, if we have a format=12 without non-BMP
-	# characters, either drop format=12 one or convert it
-	# to format=4 if there's not one.
```
Comment (anthrotype) [inline Lib/fontTools/subset/__init__.py:2231]: please keep the TODO about `convert it to format=4 if there's not one`

File: Lib/fontTools/subset/__init__.py [hunk 4]
```diff
@@ -2207,28 +2207,58 @@ def prune_pre_subset(self, font, options):
...
 			t.cmap = {u:g for u,g in t.cmap.items()
-				      if g in s.glyphs_requested}
+					  if g in s.glyphs_requested}
+
+	# Fomat 12 tables are redundant if they contain just the same BMP codepoints
+	# their little BMP-only encoding siblings contain.
+	for t in tables_format12:
+		if (
+			t.platformID == 0
+			and t.platEncID == 4
```
Comment (anthrotype) [inline Lib/fontTools/subset/__init__.py:2245]: comment what platEncID 4 and 10 mean

File: Lib/fontTools/subset/__init__.py [hunk 5]
```diff
@@ -2219,16 +2229,36 @@ def subset_glyphs(self, s):
...
 				      if g in s.glyphs_requested}
+
+	# Fomat 12 tables are redundant if they contain just the same BMP codepoints
+	# their little BMP-only encoding siblings contain.
+	for t in tables_format12:
+		if (
+			t.platformID == 0
+			and t.platEncID == 4
+			and t.language in table_plat0_enc3
+			and set(table_plat0_enc3[t.language].cmap) == set(t.cmap)
```
Comment (anthrotype) [inline Lib/fontTools/subset/__init__.py:2247]: why cast to set? can't you compare the dicts directly? or their .keys()?

File: Lib/fontTools/subset/__init__.py (hunk 1 above)
Comment (madig) [reply Lib/fontTools/subset/__init__.py]: Removed, that was unintentional.

File: Lib/fontTools/subset/__init__.py (hunk 5 above)
Comment (madig) [reply Lib/fontTools/subset/__init__.py]: Works :o

**Review by anthrotype — APPROVED** [review]
LGTM

**Pull request merged**